File: scripts/node/parsers/assignmentRulesParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

function extractAllBlocks(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'g');
  const results = [];
  let m;
  while ((m = re.exec(xml)) !== null) results.push(m[1]);
  return results;
}

/**
 * Reads Assignment Rules metadata from SFDX source format.
 * File naming: <ObjectName>.assignmentRules-meta.xml
 * @param {string} typeDir - path to the assignmentRules/ folder
 * @returns {{ rulesets: Array }}
 */
async function parseAssignmentRules(typeDir) {
  const rulesets = [];
  if (!fs.existsSync(typeDir)) return { rulesets };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.assignmentRules-meta.xml'));

  for (const file of files) {
    const xml      = fs.readFileSync(path.join(typeDir, file), 'utf8');
    const object   = file.replace('.assignmentRules-meta.xml', '');
    const blocks   = extractAllBlocks(xml, 'assignmentRule');

    const rules = blocks.map(block => ({
      name:   extractTag(block, 'fullName'),
      active: extractTag(block, 'active') === 'true',
    }));

    rulesets.push({ object, activeRules: rules.filter(r => r.active).length, totalRules: rules.length, rules });
  }

  return { rulesets };
}

module.exports = { parseAssignmentRules };

File: scripts/python/parsers/assignment_rules_parser.py
"""Parser for Assignment Rules metadata (.assignmentRules-meta.xml)."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def _extract_all_blocks(xml: str, tag: str) -> list[str]:
    return re.findall(rf'<{tag}[^>]*>[\s\S]*?</{tag}>', xml)


def parse_assignment_rules(type_dir: str) -> dict:
    """
    @param type_dir: path to the assignmentRules/ folder
    @returns: { rulesets: list }
    """
    rulesets = []
    p = Path(type_dir)
    if not p.exists():
        return {'rulesets': rulesets}

    for f in sorted(p.glob('*.assignmentRules-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        obj = f.name.replace('.assignmentRules-meta.xml', '')
        blocks = _extract_all_blocks(xml, 'assignmentRule')

        rules = [
            {
                'name':   _extract_tag(b, 'fullName'),
                'active': _extract_tag(b, 'active') == 'true',
            }
            for b in blocks
        ]

        rulesets.append({
            'object':      obj,
            'activeRules': sum(1 for r in rules if r['active']),
            'totalRules':  len(rules),
            'rules':       rules,
        })

    return {'rulesets': rulesets}
